### PR TITLE
Fix support for having multiple source modules of the same name.

### LINKED ISSUE #1187: Names from nested namespaces are not resolved correctly
If
- the same type name appears in two different nested namespaces (e.g., as `a::foo::bar` and `b::foo::bar`) and
- `a` and `b` are imported in that order

we will resolve uses of `foo::bar` in `b` to `a::foo::bar`. I suspect this is due to us not using fully qualified names when resolving names, but taking shortcuts for local names.

I am attaching [an example setup](https://github.com/zeek/spicy/files/8757625/fubar.zip) which shows this behavior.

This is a pretty nasty issue as it can lead to us silently emitting different code than expected by the author.

### PR DESCRIPTION
**Pull request by rsmmr**
This should fix #1187. The initial commits just fix smaller, independent stuff, the main commit the last one.

### CODE REVIEW

**Comment by rsmmr** [comment]
So the plugin test  fails for DHCP. I tried to build that manually and see that it fails with an import error now:

```
[error] /Users/robin/work/spicy/spicy-dhcp/analyzer/zeek_analyzer.spicy:5:1: cannot import module 'DHCP': cannot find file
```

Looking at that file, it looks like this:
```
module Zeek_DHCP;

import DHCP from protocols;
# import DHCP;
import zeek;
....
```

That `from protocols` seems odd, I wouldn't expect that to work actually. Instead, `import DHCP` should work if `analyzer.spicy` got renamed to `dhcp.spicy`. I'm wondering if this is a problem with the analyzer, and not this branch. @bbannier Am I missing something?

**Comment by bbannier** [comment]
> That `from protocols` seems odd, I wouldn't expect that to work actually. Instead, `import DHCP` should work if `analyzer.spicy` got renamed to `dhcp.spicy`. I'm wondering if this is a problem with the analyzer, and not this branch.

That from `from protocols` came from the [original version](https://github.com/zeek/spicy/blob/6b7cbc2b80d3eba83c516673dfb6bfcda0062641/zeek/plugin/lib/protocols/zeek_dhcp.spicy#L5-L7) when we had a bunch of protocol analyzers bundled with Spicy, and somehow survived the move to zeek/spicy-analyzers and then to zeek/spicy-dhcp. I'll prepare a fix for that in zeek/spicy-dhcp.

**Comment by bbannier** [comment]
zeek/spicy-dhcp is now fixed and a pushed a new tag. I retriggered the failing job; let's see if this passes 🤞.

**Comment by rsmmr** [comment]
Ok, thanks, back in my court. :-)

**Comment by rsmmr** [comment]
Test pass now, except that LGTM timed out; restarted that.

**Review by bbannier — APPROVED** [review]
I couldn't validate that this changes all uses of `id` to `uniqueID`/`scopedID`, but the test suite passes.

File: hilti/toolchain/src/compiler/unit.cc [hunk 1]
```diff
@@ -50,25 +50,27 @@ bool runHook(bool* modified, const Plugin& plugin, const Node* module, const std
...
-}
+/*
+ * Result<std::shared_ptr<Unit>> Unit::fromCache(const std::shared_ptr<Context>& context, const hilti::ID& id,
+ *                                               const hilti::rt::filesystem::path& extension) {
+ *     if ( auto cached = context->lookupUnit(id, extension) )
+ *         return cached->unit;
+ *     else
+ *         return result::Error(fmt("unknown module %s", id));
+ * }
+ */
```
[the remark this answers is not in the document]
Comment (rsmmr) [reply hilti/toolchain/src/compiler/unit.cc:68]: Yep, indeed.

**Pull request merged**